Add Kokkos profiling regions

aprokop opened the pull request:
Original discussion in ornl-cees/datatransferkit#534.

Rombur commented:
This looks nice what's left to do?

aprokop commented:
Just need to rebase and rerun to see if everything looks OK.

dalg24 commented on src/ArborX_LinearBVH.hpp:
```diff
@@ -134,6 +134,8 @@ BoundingVolumeHierarchy<DeviceType>::BoundingVolumeHierarchy(
           Kokkos::ViewAllocateWithoutInitializing( "internal_and_leaf_nodes" ),
           _size > 0 ? 2 * _size - 1 : 0 )
 {
+    Kokkos::Profiling::pushRegion( "ArborX:BVH:construction" );
```
I suppose this excludes the memory allocation for the nodes.  Please add a comment.

aprokop replied:
That's a fair point, as it implicitly assumes 0 associated cost. I think it's fine as it is, as otherwise will need to move that View creation inside the constructor. Will add a comment.

dalg24 commented on src/ArborX_LinearBVH.hpp:
```diff
@@ -147,38 +149,52 @@ BoundingVolumeHierarchy<DeviceType>::BoundingVolumeHierarchy(
     // calculate morton code of all objects
     Kokkos::View<unsigned int *, DeviceType> morton_indices(
         Kokkos::ViewAllocateWithoutInitializing( "morton" ), size() );
+    Kokkos::Profiling::pushRegion( "ArborX:BVH:assign_morton_codes" );
```
Did you mean to exclude the allocation of the array?

aprokop replied:
I don't think it matters, but I'm OK moving it up.

aprokop commented:
Force-rebased (so it could actually build and run tests), and addressed the comments.

dalg24 commented on src/details/ArborX_DetailsBoundingVolumeHierarchyImpl.hpp:
```diff
@@ -220,6 +225,9 @@ void BoundingVolumeHierarchyImpl<DeviceType>::queryDispatch(
     }
+    Kokkos::Profiling::popRegion();
+
+    Kokkos::Profiling::pushRegion( "ArborX:BVH:invalid_removal" );
```
```suggestion
    Kokkos::Profiling::pushRegion( "ArborX:BVH:filter_out_invalid_entries" );
```

dalg24 commented on src/details/ArborX_DetailsBoundingVolumeHierarchyImpl.hpp:
```diff
@@ -86,19 +86,23 @@ void BoundingVolumeHierarchyImpl<DeviceType>::queryDispatch(
     Kokkos::View<int *, DeviceType> &offset, NearestQueryAlgorithm which,
     Kokkos::View<double *, DeviceType> *distances_ptr )
 {
+    Kokkos::Profiling::pushRegion( "ArborX:BVH:knn" );
```
```suggestion
    Kokkos::Profiling::pushRegion( "ArborX:BVH:nearest_queries" );
```

dalg24 commented on src/details/ArborX_DetailsBoundingVolumeHierarchyImpl.hpp:
```diff
@@ -308,16 +319,20 @@ void BoundingVolumeHierarchyImpl<DeviceType>::queryDispatch(
     Kokkos::View<int *, DeviceType> &indices,
     Kokkos::View<int *, DeviceType> &offset, int buffer_size )
 {
+    Kokkos::Profiling::pushRegion( "ArborX:BVH:spatial" );
```
```suggestion
    Kokkos::Profiling::pushRegion( "ArborX:BVH:spatial_queries" );
```

dalg24 commented on src/details/ArborX_DetailsBoundingVolumeHierarchyImpl.hpp:
```diff
@@ -341,10 +356,11 @@ void BoundingVolumeHierarchyImpl<DeviceType>::queryDispatch(
     // [ 2 2 2 .... 2 0 ]
     //   ^            ^
     //   0th          Nth element in the view
+    Kokkos::Profiling::pushRegion( "ArborX:BVH:first_pass" );
```
Should that one include the memory allocation for the `offset` array?

aprokop replied:
Don't think so.

dalg24 commented on src/details/ArborX_DetailsBoundingVolumeHierarchyImpl.hpp:
```diff
@@ -420,6 +439,8 @@ void BoundingVolumeHierarchyImpl<DeviceType>::queryDispatch(
+        Kokkos::Profiling::popRegion();
     }
     // do not copy if by some miracle each query exactly yielded as many results
     // as the buffer size
```
Should we add a region for the copy if the buffer optimization succeeded?  (This is not a rhetorical question)

aprokop replied:
We'll have a counter there for parallel_for region anyway, so no.

dalg24 replied:
Virtually all have counters for parallel regions so I am not convinced by your argument

dalg24 commented on src/details/ArborX_DetailsBoundingVolumeHierarchyImpl.hpp:
```diff
@@ -115,6 +119,7 @@ void BoundingVolumeHierarchyImpl<DeviceType>::queryDispatch(
     reallocWithoutInitializing( indices, n_results );
     int const invalid_index = -1;
     Kokkos::deep_copy( indices, invalid_index );
```
Does that allocating memory for `offset` and finding out how much memory to allocate for `indices` before filling with the array with invalid entries belong to another region?

aprokop replied:
No region. The goal was not to have 100% coverage of everything with region, but rather cover interesting pieces.

dalg24 replied:
How were interesting pieces identified?

dalg24 commented on src/ArborX_LinearBVH.hpp:
```diff
@@ -147,38 +151,52 @@ BoundingVolumeHierarchy<DeviceType>::BoundingVolumeHierarchy(
     auto permutation_indices = Details::sortObjects( morton_indices );
+    Kokkos::Profiling::popRegion();
     Details::TreeConstruction<DeviceType>::initializeLeafNodes(
         primitives, permutation_indices, getLeafNodes() );
```
No region for that one even though we added one for the single leaf case?

aprokop commented on src/ArborX_LinearBVH.hpp:
```diff
@@ -134,6 +134,10 @@ BoundingVolumeHierarchy<DeviceType>::BoundingVolumeHierarchy(
           _size > 0 ? 2 * _size - 1 : 0 )
 {
+    // This assumes that the _internal_and_leaf_nodes view allocation takes
+    // very little time.
```
The name of the timer is "BVH construction". Thus, it is implied that this is how long it would take to construct it. If the view initialization takes too much time, it breaks this assumption.

aprokop commented:
The fact that we are arguing means that we need to define what constitutes a region, when it is appropriate to use one, and what to include/not include in it.

Rombur commented:
I think that we should try to cover 100% of the code. Otherwise we may miss something important.

aprokop commented:
We *are* covering 100% of code in overall region, the question is about nested regions. We don't really need to cover individual parallel_for, for example, as there are timers for those. And whether we should cover allocation of structures together with algorithms is up to debate.

aprokop commented:
@dalg24 @Rombur I think I addressed all your comments. I grouped together `popRegion` and `pushRegion` to indicate 100% coverage. And now most regions include the data allocations/reallocations too.
